File: jvm-experiments-commons/src/main/java/kr/kth/commons/timeperiod/timerange/CalendarTimeRange.java
package kr.kth.commons.timeperiod.timerange;

import com.google.common.base.Objects;
import kr.kth.commons.base.Guard;
import kr.kth.commons.timeperiod.*;
import kr.kth.commons.timeperiod.tools.TimeTool;
import kr.kth.commons.tools.HashTool;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;

import static kr.kth.commons.base.Guard.shouldNotBeNull;

/**
 * {@link ITimeCalendar} 기준의 기간을 나타냅니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 12. 12. 23.
 */
@Slf4j
public abstract class CalendarTimeRange extends TimeRange implements ICalendarTimeRange {

	private static final long serialVersionUID = 3653199102426865862L;

	private static TimeRange toCalendarTimeRange(ITimePeriod period, ITimePeriodMapper mapper) {
		shouldNotBeNull(period, "period");
		if (mapper == null)
			mapper = new TimeCalendar();

		DateTime mappedStart = mapper.mapStart(period.getStart());
		DateTime mappedEnd = mapper.mapEnd(period.getEnd());

		TimeTool.assertValidPeriod(mappedStart, mappedEnd);
		TimeRange mapped = new TimeRange(mappedStart, mappedEnd);

		if (log.isDebugEnabled())
			log.debug("TimeCalendar 매퍼를 이용하여 기간을 매핑했습니다. period=[{}], mapped=[{}]", period, mapped);

		return mapped;
	}

	private final ITimeCalendar timeCalendar;

	// region << constructors >>

	protected CalendarTimeRange(DateTime start, DateTime end) {
		this(start, end, new TimeCalendar());
	}

	protected CalendarTimeRange(DateTime start, DateTime end, ITimeCalendar timeCalendar) {
		this(new TimeRange(start, end), timeCalendar);
	}

	protected CalendarTimeRange(DateTime start, long duration) {
		this(new TimeRange(start, duration), new TimeCalendar());
	}

	protected CalendarTimeRange(DateTime start, long duration, ITimeCalendar timeCalendar) {
		this(new TimeRange(start, duration), timeCalendar);
	}

	protected CalendarTimeRange(ITimePeriod period) {
		this(period, new TimeCalendar());
	}

	protected CalendarTimeRange(ITimePeriod period, ITimeCalendar timeCalendar) {
		super(toCalendarTimeRange(period, timeCalendar), true);
		this.timeCalendar = timeCalendar;
	}

	// endregion

	@Override
	public ITimeCalendar getTimeCalendar() {
		return timeCalendar;
	}

	public int getStartYear() {
		return getStart().getYear();
	}

	public int getStartMonth() {
		return getStart().getMonthOfYear();
	}

	public int getStartDay() {
		return getStart().getDayOfMonth();
	}

	public int getStartHour() {
		return getStart().getHourOfDay();
	}

	public int getStartMinute() {
		return getStart().getMinuteOfHour();
	}


	public int getEndYear() {
		return getEnd().getYear();
	}

	public int getEndMonth() {
		return getEnd().getMonthOfYear();
	}

	public int getEndDay() {
		return getStart().getDayOfMonth();
	}

	public int getEndHour() {
		return getStart().getHourOfDay();
	}

	public int getEndMinute() {
		return getStart().getMinuteOfHour();
	}

	public DateTime getMappedStart() {
		return timeCalendar.mapStart(getStart());
	}

	public DateTime getMappedEnd() {
		return timeCalendar.mapEnd(getEnd());
	}

	public DateTime getUnmappedStart() {
		return timeCalendar.unmapStart(getStart());
	}

	public DateTime getUnmappedEnd() {
		return timeCalendar.unmapEnd(getEnd());
	}

	public DateTime getStartMonthStart() {
		return TimeTool.trimToDay(getStart());
	}

	public DateTime getEndMonthStart() {
		return TimeTool.trimToDay(getEnd());
	}

	public DateTime getStartDayStart() {
		return TimeTool.trimToHour(getStart());
	}

	public DateTime getEndDayEnd() {
		return TimeTool.trimToHour(getEnd());
	}

	public DateTime getStartHourStart() {
		return TimeTool.trimToMinute(getStart());
	}

	public DateTime getEndHourEnd() {
		return TimeTool.trimToMinute(getEnd());
	}

	public DateTime getStartMinuteStart() {
		return TimeTool.trimToSecond(getStart());
	}

	public DateTime getEndMinuteEnd() {
		return TimeTool.trimToSecond(getEnd());
	}

	public DateTime getStartSecondStart() {
		return TimeTool.trimToMillis(getStart());
	}

	public DateTime getEndSecondEnd() {
		return TimeTool.trimToMillis(getEnd());
	}

	@Override
	public ITimePeriod copy(long offset) {
		return toCalendarTimeRange(super.copy(offset), timeCalendar);
	}

	@Override
	protected String format(ITimeFormatter formatter) {
		return formatter.getCalendarPeriod(formatter.getDate(getStart()),
		                                   formatter.getDate(getEnd()),
		                                   getDuration());
	}

	public int compareTo(CalendarTimeRange other) {
		return getStart().compareTo(other.getStart());
	}

	@Override
	public int compareTo(ITimePeriod o) {
		Guard.shouldNotBeNull(o, "o");

		if (o instanceof CalendarTimeRange)
			return compareTo((CalendarTimeRange) o);

		return 1;
	}


	@Override
	public boolean equals(Object other) {
		return (other != null) && (other instanceof ICalendarTimeRange) && hashCode() == other.hashCode();
	}

	@Override
	public int hashCode() {
		return HashTool.compute(super.hashCode(), timeCalendar);
	}

	@Override
	public String toString() {
		return Objects.toStringHelper(this)
		              .add("description", getDescription(null))
		              .add("timeCalendar", timeCalendar)
		              .toString();
	}
}

File: jvm-experiments-timeperiod/src/main/scala/kr/kth/timeperiod/timerange/CalendarTimeRange.scala
package kr.kth.timeperiod.timerange

import com.google.common.base.Objects
import java.lang.String
import kr.kth.commons.Guard._
import kr.kth.commons.tools.ScalaHash
import kr.kth.timeperiod._
import org.joda.time.DateTime

/**
 * kr.kth.timeperiod.timerange.ICalendarTimeRange
 * User: sunghyouk.bae@gmail.com
 * Date: 13. 1. 15
 */
class CalendarTimeRange(start: DateTime, end: DateTime, val timeCalendar: ITimeCalendar)
    extends TimeRange(start, end, false) with ICalendarTimeRange {

    def this(period: ITimePeriod, timeCalendar: ITimeCalendar) {
        this(period.getStart, period.getEnd, timeCalendar)
    }

    def this(start: DateTime, duration: Long, timeCalendar: ITimeCalendar) {
        this(start, start.plus(duration), timeCalendar)
    }

    override def getTimeCalendar: ITimeCalendar = timeCalendar
}

object CalendarTimeRange {

    def apply(period: ITimePeriod): CalendarTimeRange = apply(period, TimeCalendar.Default)

    def apply(period: ITimePeriod, mapper: ITimeCalendar): CalendarTimeRange = {
        new CalendarTimeRange(toCalendarTimeRange(period, mapper), mapper)
    }

    def apply(start: DateTime, end: DateTime, timeCalendar: ITimeCalendar): CalendarTimeRange = {
        new CalendarTimeRange(start, end, timeCalendar)
    }

    def toCalendarTimeRange(period: ITimePeriod, mapper: ITimePeriodMapper = TimeCalendar.Default): TimeRange = {
        shouldNotBeNull(period, "period")
        val mappedStart: DateTime = mapper.mapStart(period.getStart)
        val mappedEnd: DateTime = mapper.mapEnd(period.getEnd)

        Times.assertValidPeriod(mappedStart, mappedEnd)
        val mapped: TimeRange = new TimeRange(mappedStart, mappedEnd)
        log.debug("TimeCalendar 매퍼를 이용하여 기간을 매핑했습니다. period=[{}], mapped=[{}]", period, mapped)

        mapped
    }
}

trait ICalendarTimeRange extends ITimeRange {

    def getTimeCalendar: ITimeCalendar

    def getStartYear: Int = getStart.getYear

    def getStartMonth: Int = getStart.getMonthOfYear

    def getStartDay: Int = getStart.getDayOfMonth

    def getStartHour: Int = getStart.getHourOfDay

    def getStartMinute: Int = getStart.getMinuteOfHour

    def getEndYear: Int = getEnd.getYear

    def getEndMonth: Int = getEnd.getMonthOfYear

    def getEndDay: Int = getStart.getDayOfMonth

    def getEndHour: Int = getStart.getHourOfDay

    def getEndMinute: Int = getStart.getMinuteOfHour

    def getMappedStart: DateTime = getTimeCalendar.mapStart(getStart)

    def getMappedEnd: DateTime = getTimeCalendar.mapEnd(getEnd)

    def getUnmappedStart: DateTime = getTimeCalendar.unmapStart(getStart)

    def getUnmappedEnd: DateTime = getTimeCalendar.unmapEnd(getEnd)

    def getStartMonthStart: DateTime = Times.trimToDay(getStart)

    def getEndMonthStart: DateTime = Times.trimToDay(getEnd)

    def getStartDayStart: DateTime = Times.trimToHour(getStart)

    def getEndDayEnd: DateTime = Times.trimToHour(getEnd)

    def getStartHourStart: DateTime = Times.trimToMinute(getStart)

    def getEndHourEnd: DateTime = Times.trimToMinute(getEnd)

    def getStartMinuteStart: DateTime = Times.trimToSecond(getStart)

    def getEndMinuteEnd: DateTime = Times.trimToSecond(getEnd)

    def getStartSecondStart: DateTime = Times.trimToMillis(getStart)

    def getEndSecondEnd: DateTime = Times.trimToMillis(getEnd)

    override def copy(offset: Long): ITimePeriod = {
        CalendarTimeRange.toCalendarTimeRange(super.copy(offset), getTimeCalendar)
    }

    override protected def format(formatter: Option[ITimeFormatter]): String = {
        val fmt = formatter.getOrElse(TimeFormatter.instance)
        fmt.getCalendarPeriod(fmt.getDate(getStart), fmt.getDate(getEnd), getDuration)
    }

    override def hashCode: Int =
        ScalaHash.compute(super.hashCode, getTimeCalendar)

    protected override def buildStringHelper(): Objects.ToStringHelper = {
        super.buildStringHelper()
        .add("timeCalendar", getTimeCalendar)
    }
}

